File: src/main/java/com/healthpal/api/service/UserService.java
package com.healthpal.api.service;

import com.healthpal.api.model.User;
import com.healthpal.api.model.Role;
import com.healthpal.api.repository.UserRepository;
import com.healthpal.api.repository.RoleRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * UserService - Business Logic Layer for User Management
 * 
 * This service is part of the MVC Layered Architecture (Service/Business Logic Layer)
 * It handles:
 * - User profile management
 * - User role assignment
 * - User activation/deactivation
 * - User data retrieval
 * 
 * @author HealthPal Team - Person A (Authentication + Admin + Core Architecture)
 * @version 1.0
 */
@Slf4j
@Service
@Transactional
public class UserService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private RoleRepository roleRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    /**
     * Get user by ID
     * @param userId The ID of the user
     * @return Optional containing the user if found
     */
    public Optional<User> getUserById(Long userId) {
        log.debug("Fetching user with ID: {}", userId);
        return userRepository.findById(userId);
    }

    /**
     * Get user by username
     * @param username The username
     * @return Optional containing the user if found
     */
    public Optional<User> getUserByUsername(String username) {
        log.debug("Fetching user with username: {}", username);
        return userRepository.findByUsername(username);
    }

    /**
     * Get user by email
     * @param email The email
     * @return Optional containing the user if found
     */
    public Optional<User> getUserByEmail(String email) {
        log.debug("Fetching user with email: {}", email);
        return userRepository.findByEmail(email);
    }

    /**
     * Get all active users
     * @return List of all active users
     */
    public List<User> getAllActiveUsers() {
        log.debug("Fetching all active users");
        return userRepository.findAllActiveUsers();
    }

    /**
     * Get all admin users
     * @return List of all admin users
     */
    public List<User> getAllAdmins() {
        log.debug("Fetching all admin users");
        return userRepository.findAllAdmins();
    }

    /**
     * Get users by role
     * @param roleName The name of the role
     * @return List of users with the specified role
     */
    public List<User> getUsersByRole(String roleName) {
        log.debug("Fetching users with role: {}", roleName);
        return userRepository.findUsersByRole(roleName);
    }

    /**
     * Update user profile
     * @param userId The ID of the user to update
     * @param user The updated user information
     * @return Updated user
     * @throws RuntimeException if user not found
     */
    public User updateUserProfile(Long userId, User user) {
        log.info("Updating user profile for user ID: {}", userId);

        User existingUser = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));

        if (user.getFirstName() != null) {
            existingUser.setFirstName(user.getFirstName());
        }
        if (user.getLastName() != null) {
            existingUser.setLastName(user.getLastName());
        }
        if (user.getPhoneNumber() != null) {
            existingUser.setPhoneNumber(user.getPhoneNumber());
        }
        if (user.getProfilePicture() != null) {
            existingUser.setProfilePicture(user.getProfilePicture());
        }

        return userRepository.save(existingUser);
    }

    /**
     * Change user password
     * @param userId The ID of the user
     * @param oldPassword The old password
     * @param newPassword The new password
     * @return Updated user
     * @throws RuntimeException if user not found or old password is incorrect
     */
    public User changePassword(Long userId, String oldPassword, String newPassword) {
        log.info("Changing password for user ID: {}", userId);

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));

        if (!passwordEncoder.matches(oldPassword, user.getPassword())) {
            log.warn("Invalid old password for user ID: {}", userId);
            throw new RuntimeException("Old password is incorrect");
        }

        user.setPassword(passwordEncoder.encode(newPassword));
        return userRepository.save(user);
    }

    /**
     * Assign role to user
     * @param userId The ID of the user
     * @param roleName The name of the role to assign
     * @return Updated user
     * @throws RuntimeException if user or role not found
     */
    public User assignRoleToUser(Long userId, String roleName) {
        log.info("Assigning role {} to user ID: {}", roleName, userId);

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));

        Role role = roleRepository.findByName(roleName)
                .orElseThrow(() -> new RuntimeException("Role not found: " + roleName));

        user.addRole(role);
        return userRepository.save(user);
    }

    /**
     * Remove role from user
     * @param userId The ID of the user
     * @param roleName The name of the role to remove
     * @return Updated user
     * @throws RuntimeException if user or role not found
     */
    public User removeRoleFromUser(Long userId, String roleName) {
        log.info("Removing role {} from user ID: {}", roleName, userId);

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));

        Role role = roleRepository.findByName(roleName)
                .orElseThrow(() -> new RuntimeException("Role not found: " + roleName));

        user.removeRole(role);
        return userRepository.save(user);
    }

    /**
     * Activate user account
     * @param userId The ID of the user to activate
     * @return Updated user
     * @throws RuntimeException if user not found
     */
    public User activateUser(Long userId) {
        log.info("Activating user ID: {}", userId);

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));

        user.setIsActive(true);
        return userRepository.save(user);
    }

    /**
     * Deactivate user account
     * @param userId The ID of the user to deactivate
     * @return Updated user
     * @throws RuntimeException if user not found
     */
    public User deactivateUser(Long userId) {
        log.info("Deactivating user ID: {}", userId);

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));

        user.setIsActive(false);
        return userRepository.save(user);
    }

    /**
     * Verify user email
     * @param userId The ID of the user
     * @return Updated user
     * @throws RuntimeException if user not found
     */
    public User verifyUserEmail(Long userId) {
        log.info("Verifying email for user ID: {}", userId);

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));

        user.setIsEmailVerified(true);
        return userRepository.save(user);
    }

    /**
     * Delete user
     * @param userId The ID of the user to delete
     * @throws RuntimeException if user not found
     */
    public void deleteUser(Long userId) {
        log.info("Deleting user ID: {}", userId);

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userId));

        userRepository.delete(user);
    }
}

File: service/userService.js
const userRepository = require('../repositories/userRepository');
const roleRepository = require('../repositories/roleRepository');
const bcrypt = require('bcryptjs');
const logger = require('../utils/logger');

class UserService {
    async getUserById(userId) {
        try {
            logger.debug(`Fetching user with ID: ${userId}`);
            const user = await userRepository.findById(userId);
            if (!user) {
                logger.debug(`User not found with ID: ${userId}`);
                return null;
            }
            return user;
        } catch (error) {
            logger.error(`Error in getUserById: ${error.message}`);
            throw error;
        }
    }

    async getUserByUsername(username) {
        try {
            logger.debug(`Fetching user with username: ${username}`);
            const user = await userRepository.findByUsername(username);
            if (!user) {
                logger.debug(`User not found with username: ${username}`);
                return null;
            }
            return user;
        } catch (error) {
            logger.error(`Error in getUserByUsername: ${error.message}`);
            throw error;
        }
    }

    async getUserByEmail(email) {
        try {
            logger.debug(`Fetching user with email: ${email}`);
            const user = await userRepository.findByEmail(email);
            if (!user) {
                logger.debug(`User not found with email: ${email}`);
                return null;
            }
            return user;
        } catch (error) {
            logger.error(`Error in getUserByEmail: ${error.message}`);
            throw error;
        }
    }

    async getAllActiveUsers() {
        try {
            logger.debug('Fetching all active users');
            return await userRepository.findAllActiveUsers();
        } catch (error) {
            logger.error(`Error in getAllActiveUsers: ${error.message}`);
            throw error;
        }
    }

    async getAllAdmins() {
        try {
            logger.debug('Fetching all admin users');
            return await userRepository.findAllAdmins();
        } catch (error) {
            logger.error(`Error in getAllAdmins: ${error.message}`);
            throw error;
        }
    }

    async getUsersByRole(roleName) {
        try {
            logger.debug(`Fetching users with role: ${roleName}`);
            return await userRepository.findUsersByRole(roleName);
        } catch (error) {
            logger.error(`Error in getUsersByRole: ${error.message}`);
            throw error;
        }
    }

    async updateUserProfile(userId, userData) {
        try {
            logger.info(`Updating user profile for user ID: ${userId}`);
            const existingUser = await userRepository.findById(userId);
            if (!existingUser) {
                throw new Error(`User not found with ID: ${userId}`);
            }

            const updates = {};
            if (userData.firstName !== undefined) {
                updates.firstName = userData.firstName;
            }
            if (userData.lastName !== undefined) {
                updates.lastName = userData.lastName;
            }
            if (userData.phoneNumber !== undefined) {
                updates.phoneNumber = userData.phoneNumber;
            }
            if (userData.profilePicture !== undefined) {
                updates.profilePicture = userData.profilePicture;
            }

            if (Object.keys(updates).length === 0) {
                logger.warn(`No valid fields to update for user ID: ${userId}`);
                return existingUser;
            }

            const updatedUser = await userRepository.update(userId, updates);
            logger.info(`User profile updated successfully for user ID: ${userId}`);
            return updatedUser;
        } catch (error) {
            logger.error(`Error updating user profile: ${error.message}`);
            throw error;
        }
    }

    async changePassword(userId, oldPassword, newPassword) {
        try {
            logger.info(`Changing password for user ID: ${userId}`);
            const user = await userRepository.findById(userId);
            if (!user) {
                throw new Error(`User not found with ID: ${userId}`);
            }

            const userWithPassword = await userRepository.findByUsername(user.username);
            if (!userWithPassword) {
                throw new Error(`User not found with username: ${user.username}`);
            }

            const isPasswordValid = await bcrypt.compare(oldPassword, userWithPassword.password);
            if (!isPasswordValid) {
                logger.warn(`Invalid old password for user ID: ${userId}`);
                throw new Error('Old password is incorrect');
            }

            const hashedPassword = await bcrypt.hash(newPassword, 10);
            const updatedUser = await userRepository.update(userId, { password: hashedPassword });
            logger.info(`Password changed successfully for user ID: ${userId}`);
            return updatedUser;
        } catch (error) {
            logger.error(`Error changing password: ${error.message}`);
            throw error;
        }
    }

    async assignRoleToUser(userId, roleName) {
        try {
            logger.info(`Assigning role ${roleName} to user ID: ${userId}`);
            const [user, role] = await Promise.all([
                userRepository.findById(userId),
                roleRepository.findByName(roleName)
            ]);

            if (!user) {
                throw new Error(`User not found with ID: ${userId}`);
            }
            if (!role) {
                throw new Error(`Role not found: ${roleName}`);
            }

            const hasRole = user.roles.some(r => r._id.toString() === role._id.toString());
            if (!hasRole) {
                await user.addRole(role);
                logger.info(`Role ${roleName} assigned to user ID: ${userId}`);
            } else {
                logger.debug(`User already has role ${roleName}`);
            }

            return await userRepository.findById(userId);
        } catch (error) {
            logger.error(`Error assigning role: ${error.message}`);
            throw error;
        }
    }

    async removeRoleFromUser(userId, roleName) {
        try {
            logger.info(`Removing role ${roleName} from user ID: ${userId}`);
            const [user, role] = await Promise.all([
                userRepository.findById(userId),
                roleRepository.findByName(roleName)
            ]);

            if (!user) {
                throw new Error(`User not found with ID: ${userId}`);
            }
            if (!role) {
                throw new Error(`Role not found: ${roleName}`);
            }

            const hasRole = user.roles.some(r => r._id.toString() === role._id.toString());
            if (hasRole) {
                await user.removeRole(role);
                logger.info(`Role ${roleName} removed from user ID: ${userId}`);
            } else {
                logger.debug(`User does not have role ${roleName}`);
            }

            return await userRepository.findById(userId);
        } catch (error) {
            logger.error(`Error removing role: ${error.message}`);
            throw error;
        }
    }

    async activateUser(userId) {
        try {
            logger.info(`Activating user ID: ${userId}`);
            const user = await userRepository.findById(userId);
            if (!user) {
                throw new Error(`User not found with ID: ${userId}`);
            }

            const updatedUser = await userRepository.update(userId, { isActive: true });
            logger.info(`User activated successfully: ${userId}`);
            return updatedUser;
        } catch (error) {
            logger.error(`Error activating user: ${error.message}`);
            throw error;
        }
    }

    async deactivateUser(userId) {
        try {
            logger.info(`Deactivating user ID: ${userId}`);
            const user = await userRepository.findById(userId);
            if (!user) {
                throw new Error(`User not found with ID: ${userId}`);
            }

            const updatedUser = await userRepository.update(userId, { isActive: false });
            logger.info(`User deactivated successfully: ${userId}`);
            return updatedUser;
        } catch (error) {
            logger.error(`Error deactivating user: ${error.message}`);
            throw error;
        }
    }

    async verifyUserEmail(userId) {
        try {
            logger.info(`Verifying email for user ID: ${userId}`);
            const user = await userRepository.findById(userId);
            if (!user) {
                throw new Error(`User not found with ID: ${userId}`);
            }

            const updatedUser = await userRepository.update(userId, { isEmailVerified: true });
            logger.info(`Email verified successfully for user ID: ${userId}`);
            return updatedUser;
        } catch (error) {
            logger.error(`Error verifying email: ${error.message}`);
            throw error;
        }
    }

    async deleteUser(userId) {
        try {
            logger.info(`Deleting user ID: ${userId}`);
            const user = await userRepository.findById(userId);
            if (!user) {
                throw new Error(`User not found with ID: ${userId}`);
            }

            const deleted = await userRepository.delete(userId);
            if (!deleted) {
                throw new Error(`Failed to delete user with ID: ${userId}`);
            }
            logger.info(`User deleted successfully: ${userId}`);
        } catch (error) {
            logger.error(`Error deleting user: ${error.message}`);
            throw error;
        }
    }

    async getUserByUsernameOrEmail(identifier) {
        try {
            logger.debug(`Loading user details for: ${identifier}`);
            const user = await userRepository.findByUsernameOrEmail(identifier, identifier);
            if (!user) {
                logger.warn(`User not found: ${identifier}`);
                throw new Error(`User not found with identifier: ${identifier}`);
            }

            if (!user.isActive) {
                logger.warn(`User account is inactive: ${identifier}`);
                throw new Error('User account is inactive');
            }

            return user;
        } catch (error) {
            logger.error(`Error in getUserByUsernameOrEmail: ${error.message}`);
            throw error;
        }
    }

    async createUser(userData) {
        try {
            if (userData.password) {
                userData.password = await bcrypt.hash(userData.password, 10);
            }

            const user = await userRepository.save(userData);
            logger.info(`User created successfully: ${user.username}`);
            return user;
        } catch (error) {
            logger.error(`Error in createUser: ${error.message}`);
            throw error;
        }
    }

    async updateLastLogin(userId) {
        try {
            return await userRepository.update(userId, { lastLoginAt: new Date() });
        } catch (error) {
            logger.error(`Error in updateLastLogin: ${error.message}`);
            throw error;
        }
    }
}

module.exports = new UserService();